src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.hpp

```diff
@@ -0,0 +1,164 @@
+#include "jfr/utilities/jfrTypes.hpp"
+
+class JavaThread;
+class NonJavaThread;
+
+#if defined(LINUX)
+
+#include "memory/padded.hpp"
+#include "jfr/periodic/sampling/jfrSampleRequest.hpp"
+
+struct JfrCPUTimeSampleRequest {
+  JfrSampleRequest _request;
+  Tickspan _cpu_time_period;
+
+  JfrCPUTimeSampleRequest() {}
+};
+
+// Fixed size async-signal-safe SPSC linear queue backed by an array.
+// Designed to be only used under lock and read linearly
+class JfrCPUTimeTraceQueue {
+
+  // the default queue capacity, scaled if the sampling period is smaller than 10ms
+  // when the thread is started
+  static const u4 CPU_TIME_QUEUE_CAPACITY = 500;
+
+  JfrCPUTimeSampleRequest* _data;
+  u4 _capacity;
+  // next unfilled index
+  volatile u4 _head;
+
+  volatile s4 _lost_samples;
+
+  // index of the first request of the thread in native state
+  // -1 if no native request is in the queue
+  volatile s4 _first_native_request_index;
+
+public:
+  JfrCPUTimeTraceQueue(u4 capacity);
+
+  ~JfrCPUTimeTraceQueue();
+
+  // signal safe, but can't be interleaved with dequeue
+  bool enqueue(JfrCPUTimeSampleRequest& trace);
+
+  JfrCPUTimeSampleRequest& at(u4 index);
+
+  u4 size() const;
+
+  void set_size(u4 size);
+
+  u4 capacity() const;
+
+  // deletes all samples in the queue
+  void set_capacity(u4 capacity);
+
+  bool is_full() const;
+
+  bool is_empty() const;
+
+  s4 lost_samples() const;
+
+  void increment_lost_samples();
+
+  // returns the previous lost samples count
+  u4 get_and_reset_lost_samples();
+
+  void ensure_capacity(u4 capacity);
+
+  void ensure_capacity_for_period(u4 period_millis);
+
+  void clear();
+
+  void set_first_native_request_index(s4 index);
+
+  s4 first_native_request_index() const;
+};
+
+
+class JfrCPUTimeThreadSampler;
+
+class JfrCPUTimeThreadSampling : public JfrCHeapObj {
+  friend class JfrRecorder;
+ private:
+
+  JfrCPUTimeThreadSampler* _sampler;
+
+  void create_sampler(double rate, bool autoadapt);
+  void set_rate_value(double rate, bool autoadapt);
+
+  JfrCPUTimeThreadSampling();
+  ~JfrCPUTimeThreadSampling();
+
+  static JfrCPUTimeThreadSampling& instance();
+  static JfrCPUTimeThreadSampling* create();
+  static void destroy();
+
+  void update_run_state(double rate, bool autoadapt);
+
+ public:
+  static void set_rate(double rate, bool autoadapt);
+
+  static void on_javathread_create(JavaThread* thread);
+  static void on_javathread_terminate(JavaThread* thread);
+  void handle_timer_signal(siginfo_t* info, void* context);
+
+  static void send_empty_event(const JfrTicks& start_time, traceid tid, Tickspan cpu_time_period);
+  static void send_event(const JfrTicks& start_time, traceid sid, traceid tid, Tickspan cpu_time_period, bool biased);
+  static void send_lost_event(const JfrTicks& time, traceid tid, s4 lost_samples);
+
+  // Trigger sampling while a thread is not in a safepoint, from a seperate thread
+  static void trigger_is_thread_in_native_stackwalking();
```

Comment:
Is it sampling that is triggered? Sampling refers to the asynchronous signal received from the OS.

You are asking for the sampler thread to process JFR Sample Requests, right? That is, already taken samples.